yt/yt/server/http_proxy/component_discovery.cpp

```diff
@@ -0,0 +1,386 @@
+#include "component_discovery.h"
+#include "coordinator.h"
+
+#include <yt/yt/client/api/client.h>
+
+#include <util/string/split.h>
+
+namespace NYT::NHttpProxy {
+
+using namespace NApi;
+using namespace NConcurrency;
+using namespace NYTree;
+using namespace NYson;
+
+////////////////////////////////////////////////////////////////////////////////
+
+static const auto& Logger = HttpProxyLogger;
+
+////////////////////////////////////////////////////////////////////////////////
+
+namespace {
+
+////////////////////////////////////////////////////////////////////////////////
+
+void FillMasterReadOptions(TMasterReadOptions& options, const TMasterReadOptions& value)
+{
+    options = value;
+}
+
+////////////////////////////////////////////////////////////////////////////////
+
+} // namespace
+
+////////////////////////////////////////////////////////////////////////////////
+
+void Serialize(const TClusterComponentInstance& instance, IYsonConsumer* consumer)
+{
+    BuildYsonFluently(consumer)
+        .BeginMap()
+            .Item("address").Value(instance.Address)
+            .Item("type").Value(instance.Type)
+            .DoIf(instance.Error.IsOK(), [&] (auto fluent) {
+                fluent
+                    .Item("version").Value(instance.Version)
+                    .Item("start_time").Value(instance.StartTime)
+                    .Item("banned").Value(instance.Banned);
+
+                if (!instance.State.empty()) {
+                    fluent.Item("state").Value(instance.State);
+                }
+            })
+            .DoIf(!instance.Error.IsOK(), [&] (auto fluent) {
+                fluent.Item("error").Value(instance.Error);
+            })
+        .EndMap();
+}
+
+////////////////////////////////////////////////////////////////////////////////
+
+TComponentDiscoverer::TComponentDiscoverer(
+    IClientPtr client,
+    TMasterReadOptions masterReadOptions,
+    TDuration proxyDeathAge)
+    : Client_(std::move(client))
+    , MasterReadOptions_(std::move(masterReadOptions))
+    , ProxyDeathAge_(proxyDeathAge)
+{ }
+
+std::vector<TClusterComponentInstance> TComponentDiscoverer::ListNodes(EClusterComponentType component) const
+{
+    TListNodeOptions options;
+    FillMasterReadOptions(options, MasterReadOptions_);
+    options.Attributes = {
+        "register_time",
+        "version",
+        "banned",
+        "state",
+        "job_proxy_build_version",
+    };
+
+    auto rsp = WaitFor(Client_->ListNode(GetCypressDirectory(component), options))
+        .ValueOrThrow();
+    auto rspList = ConvertToNode(rsp)->AsList();
+
+    std::vector<TClusterComponentInstance> instances;
+    instances.reserve(rspList->GetChildren().size());
+
+    for (const auto& node : rspList->GetChildren()) {
+        auto version = node->Attributes().Find<TString>("version");
+        auto nodeState = node->Attributes().Get<TString>("node_state", /*defaultValue*/ "");
```

Comment:
The correct attribute name is "state", I don't know what got into me and decided to change it.